### PR TITLE
Send notifications to partecipants in issue comments

### LINKED ISSUE #1216: Participants of an issue or pull request should receive email notifications
Issue migrated from Gogs: https://github.com/gogits/gogs/issues/2929

Supposedly fixed in Gogs 6 hours ago: https://github.com/gogits/gogs/commit/23da90e25d610e94a50dd27223baaa8834f61607

To be ported over.

\cc @andreynering

### PR DESCRIPTION
**Pull request by strk**
Closes #1216

NOTE: I didn't test this with GUI notifications so might need a review
from @andreynering to deal with that -- this is just a straight port
from the Gogs commit

### CODE REVIEW

File: models/issue_mail.go [hunk 1]
```diff
@@ -48,6 +54,16 @@ func mailIssueCommentToParticipants(issue *Issue, doer *User, mentions []string)
 		tos = append(tos, to.Email)
 		names = append(names, to.Name)
 	}
+	for i := range participants {
```
Comment (lunny) [inline models/issue_mail.go:63]: ignore self on watchers too?

**Review by lunny — CHANGES_REQUESTED** [review]

**Comment by strk** [comment]
It's done one line below:

```
  if participants[i].ID == doer.ID {
    continue
```

**Comment by lunny** [comment]
I mean `watchers` not `participants`

**Comment by strk** [comment]
Done also for `watchers`:
```
        for i := range watchers {
                if watchers[i].UserID == doer.ID {
                        continue
                }
```

**Comment by lunny** [comment]
Where is the commit?

**Comment by strk** [comment]
It was already in the file, so not changed by the commit.
The code can be seen in the file as modified by the commit:
https://github.com/strk/gitea/blob/809ecfddbb8fd69b67144fcf77abb512ec12e5e8/models/issue_mail.go#L41-L44

File: models/issue_mail.go (hunk 1 above)
Comment (strk) [reply models/issue_mail.go:63]: this is already done in the file, a few lines above (see the full file)

**Comment by puffybsd** [comment]
Did some testing on 'issue notifications'.  For the most part, it worked, although I wasn't able to reproduce an issue on master, some emails were delayed and issue counts for the issue creator were out of sync (but eventually synchronized).

Using two accounts & two email accounts (one gmail, one yahoo, sparkpost relay), creator of issue is a participant, repo owner is a participant. confirmed that emails are sent to the other participant when a change is made (creator opens an issue- email sent to repo owner, not to creator; repo owner comments, email sent to creator, not to repo owner).  Let me know if there's something else you'd like tested.

**Comment by strk** [comment]
@puffybsd I'd love you to try the same on try.gitea.io to spot the difference, but I'm not sure email notification there work at the moment (due to #1223)

**Comment by lunny** [comment]
LGTM, I just made some simple tests, it works.

**Review by lunny — APPROVED** [review]

**Comment by lunny** [comment]
But maybe you didn't consider one situation. Issue creators seem will not receive emails? Except this.

**Comment by strk** [comment]
Had you tried that @lunny ? @puffybsd had you (making sure creator was not "Watching" the repo)

**Comment by lunny** [comment]
line 86 should has a bug. the poster maybe the comment poster not the issue poster. So poster should be a param to `MailParticipants`.
```Go
func (issue *Issue) MailParticipants() (err error) {
```

**Comment by strk** [comment]
@lunny actually I don't understand the issue. Are you saying that the issue poster should *not* receive an email notification when someone comments on the issue ?

File: models/issue.go [hunk 2]
```diff
@@ -1133,6 +1133,23 @@ func Issues(opts *IssuesOptions) ([]*Issue, error) {
 	return issues, nil
 }
 
+// GetParticipantsByIssueID returns all users who are participated in comments of an issue.
+func GetParticipantsByIssueID(issueID int64) ([]*User, error) {
```
Comment (ethantkoenig) [inline models/issue.go:1138]: Could we add unit tests for this new function?

**Review by ethantkoenig — CHANGES_REQUESTED** [review]

File: models/issue.go (hunk 2 above)
Comment (strk) [reply models/issue.go:1138]: Will look into it as time permits (possibly next week)

**Comment by strk** [comment]
@lunny @Unkwon actually confirmed the bug and reports it as fixed by https://github.com/gogits/gogs/commit/5f058c3f071b5a754bdd5b556b49d678bf1477f1

Great reviews, thanks everyone !

**Comment by andreynering** [comment]
Trusted LGTM

**Comment by lunny** [comment]
@strk I don't think that will fix the problem entirely, I will test this tomorrow again.

**Comment by strk** [comment]
192ca213f08b4d613c45bce22b78d7497f342ff5 stubs an initial test, and fails revealing still a bug !
(the one you're talking about @lunny ?)

I don't know what @Unknwon idea was for that function (GetParticipantsByIssueID), but  to me it does make sense for the original poster to be among the partecipants...

**Comment by strk** [comment]
So basically what's the failing test telling us is that the current `GetParticipantsByIssueID` function (in this PR) is returning all the users who commented and those who applied labels, but not the issue author. I don't know what @Unknwon thinks about this but I would not include labels changers while I would include the issue author. What do other think about it ? In other words, what consistitutes "partecipation" in an issue ?

File: models/issue.go [hunk 3]
```diff
@@ -1134,6 +1134,24 @@ func Issues(opts *IssuesOptions) ([]*Issue, error) {
...
+	userIDs := make([]int64, 0, 5)
```
[the remark this answers is not in the document]
Comment (lunny) [reply models/issue.go:1139]: nothing.

**Comment by lunny** [comment]
let L-G-T-M work

**Pull request merged**